#!/bin/bash

PATCH="baselayout-usr-merge.patch"
rm -f sys-apps/baselayout/baselayout-2.5*.ebuild || exit 1
for i in sys-apps/baselayout/baselayout-2*.ebuild; do
	cat > ${PATCH} << EOF
--- a/${i}	2013-07-28 01:01:32.000000000 +0200
+++ b/${i}	2013-10-03 19:18:09.076757264 +0200
@@ -124,6 +124,52 @@
 	done
 }
 
+usrmerge_layout() {
+	[[ -e \${EROOT}usr ]] || return
+	[[ \${EROOT}. -ef \${EROOT}usr/. ]] && return
+	[[ -d \${EROOT}usr ]] || die "\${EROOT}usr is not a directory"
+	local d
+	for d in bin sbin \$(get_all_libdirs) ; do
+		if [[ \${EROOT}\${d}/. -ef \${EROOT}usr/\${d}/. ]] ; then
+			continue
+		elif [[ -d \${EROOT}\${d} ]] ; then
+			if [[ -h \${EROOT}\${d} ]] ; then
+				eerror "\${EROOT}\${d} is a symlink but it doesn't point to usr/\${d}"
+				continue
+			elif [[ -h \${EROOT}usr/\${d} ]] ; then
+				if [[ ! -d \${EROOT}usr/\${d} ]] ; then
+					die "\${EROOT}usr/\${d} is not a directory"
+				elif [[ \${SYMLINK_LIB} = "yes" && \${d} == "lib" ]] ; then
+					[[ \${MERGE_USR} == "no" ]] && continue
+					mv "\${EROOT}\${d}"/* "\${EROOT}usr/\${d}"
+					rmdir "\${EROOT}\${d}"
+				else
+					eerror "\${EROOT}usr/\${d} is a symlink when it shouldn't be"
+					continue
+				fi
+			elif [[ -d \${EROOT}usr/\${d} ]] ; then
+				[[ \${MERGE_USR} == "no" ]] && continue
+				# it is pointless to check for duplicate files
+				# if they don't exist everything is fine
+				# if they do they will reappear on first update
+				mv "\${EROOT}\${d}"/* "\${EROOT}usr/\${d}"
+				rmdir "\${EROOT}\${d}"
+			elif [[ -e \${EROOT}usr/\${d} ]] ; then
+				die "\${EROOT}usr/\${d} is not a directory"
+			else
+				mv "\${EROOT}\${d}" "\${EROOT}usr"
+			fi
+		elif [[ -e \${EROOT}\${d} ]] ; then
+			die "\${EROOT}\${d} is not a directory"
+		fi
+		if [[ -e \${EROOT}\${d} ]] && ! rmdir "\${EROOT}\${d}" ; then
+			die "failed to recreate \${EROOT}\${d} as a symlink"
+		fi
+		ln -s usr/\${d} "\${EROOT}\${d}"
+		mkdir -p "\${EROOT}usr/\${d}"
+	done
+}
+
 pkg_preinst() {
 	# Bug #217848 - Since the remap_dns_vars() called by pkg_preinst() of
 	# the baselayout-1.x ebuild copies all the real configs from the user's
@@ -141,4 +169,5 @@
 	# stages, but they cannot be in CONTENTS.
 	# Also, we cannot reference \$S as binpkg will break so we do this.
 	multilib_layout
+	usrmerge_layout
 	if use build ; then
EOF
	git apply ${PATCH} || exit 1
	sed -i 's|,\(usr/,usr/local/\)|\1|' $i
	rm ${PATCH}
done
